game-core/src/main/java/games/strategy/triplea/image/ImageColorizer.java

```diff
@@ -0,0 +1,93 @@
+package games.strategy.triplea.image;
+
+import java.awt.Color;
+import java.awt.image.BufferedImage;
+
+/**
+ * Class used to colorize images. Based on finding the luminance from the original image then
+ * applying hue, saturation, and brightness to colorize it.
+ */
+public class ImageColorizer {
+
+  private static final int MAX_COLOR = 256;
+  private static final float LUMINANCE_RED = 0.2126f;
+  private static final float LUMINANCE_GREEN = 0.7152f;
+  private static final float LUMINANCE_BLUE = 0.0722f;
+
+  private ImageColorizer() {}
+
+  /**
+   * Apply color to the given image. This takes the hue and saturation of the given color and
+   * applies it to the image. It ignores the brightness of the color which can cause images to
+   * become poor quality.
+   */
+  public static void applyColor(final Color color, final BufferedImage image) {
+    final float[] hsb = Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), null);
+    applyHsb((int) (hsb[0] * 360), (int) (hsb[1] * 100), 0, image);
+  }
+
+  /**
+   * Apply HSB values to the given image.
+   *
+   * @param hue 0 to 360
+   * @param saturation 0 to 100
+   * @param brightness -100 to 100, which adjusts the given image's brightness
+   * @param image the image to transform
+   */
+  public static void applyHsb(
+      final int hue, final int saturation, final int brightness, final BufferedImage image) {
+
+    // Create color lookup table for luminance values
+    final int[] redLookup = new int[MAX_COLOR];
```

Comment:
Do the lookup tables save that much computation? Looking at java.awt.Color, it looks like we might be saving a few floating point operations. I wonder if there is any real performance gain here. I would suspect if there were a caching gain to be had, it would be by caching input color to final color.

Reply:
Some but didn't test performance improvement. For large images this could make a noticeable difference (think 1000x1000 pixels) as anything in double for loop is costly. For unit images they tend to be around 50x50 so not a huge deal but I think its better to just leave this since its a generic image transformation class.

Reply:
Interesting performance results.

Original: 1.8ms
No-lookup table: 1.7ms
Caching the before to after color: 2.1ms


The code with the cached lookup table is below (the no-lookup table is the same but without the hashmap):

```
  private static final Map<Color, Color> lookupTable = new HashMap<>();
  private static void colorize(
      final int hue, final int saturation, final int brightness, final BufferedImage image) {

    // Create color lookup table for luminance values
    final float hueFloat = hue / 360f;
    final float saturationFloat = saturation / 100f;

    // Loop through image to colorize each pixel
    for (int x = 0; x < image.getWidth(); x++) {
      for (int y = 0; y < image.getHeight(); y++) {
        final Color color = new Color(image.getRGB(x, y), true);

        // Don't worry about colorization if pixel is transparent
        if (color.getAlpha() == 0) {
          continue;
        }

        if(lookupTable.containsKey(color)) {
          image.setRGB(x, y, lookupTable.get(color).getRGB());
          continue;
        }

        // Avoid colorization if pixel is close to black so image stays sharp
        final float[] hsb = Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), null);
        if (hsb[2] < 0.1) {
          continue;
        }

        // Find luminance and use lookup table to set resulting color
        final int lum = findLuminance(color, brightness);
        final Color finalColor = computeColor(hueFloat, saturationFloat, lum);
        lookupTable.put(color, finalColor);
//            new Color(redLookup[lum], greenLookup[lum], blueLookup[lum], color.getAlpha());
        image.setRGB(x, y, finalColor.getRGB());
      }
    }
  }


  private static Color computeColor(final float hueFloat, float saturationFloat, int lum) {
    int i = lum;
//    for (int i = 0; i < MAX_COLOR; i++) {
      final float brightnessFloat = (float) i / (MAX_COLOR - 1);
      return Color.getHSBColor(hueFloat, saturationFloat, brightnessFloat);
  }
```

It's very surprising that the array lookup is *slower*

Reply:
Well its not much different either way with those results. I would be interested for larger images if you see the same results.